### PR TITLE
Immediately cancel grace membership

### PR DESCRIPTION
**Pull request by caillou**

### CODE REVIEW

File: servers/republik/modules/crowdfundings/lib/generateMemberships.js [hunk 1]
```diff
@@ -260,3 +262,11 @@ module.exports = async (pledgeId, pgdb, t, redis) => {
   const cache = createCache({ prefix: `User:${user.id}` }, { redis })
   cache.invalidate()
 }
+
+async function isMembershipInGracePeriod (membership, pgdb) {
+  const hasActivePeriod = await pgdb.public.membershipPeriods.count({
+    membershipId: membership.id,
+    'endDate >': new Date()
+  })
+  return !hasActivePeriod
```
Comment (patte) [inline servers/republik/modules/crowdfundings/lib/generateMemberships.js:271]:
```suggestion
  const hasPeriodEndingInTheFuture = await pgdb.public.membershipPeriods.count({
    membershipId: membership.id,
    'endDate >': new Date()
  })
  return !hasPeriodEndingInTheFuture
```

**Comment by tpreusse** [comment]
One input: wouldn't it be better if immediately would always be set to true in the cancel method?
https://github.com/orbiting/backends/blob/54025684ae123153feb4cc36ce21209c36bc284f/servers/republik/modules/crowdfundings/lib/cancelMembership.js#L40-L47

For example when a user cancels instead of updating the credit card. Currently the frontend never sends `immediately`.

**Comment by patte** [comment]
@tpreusse I'm not sure I understand. Immediately cancelling a membership with a period ending in the future leads to loosing the days between `now` and then. Which is not what most customers want. Can you explain what you mean?

**Comment by caillou** [comment]
@tpreusse could we then remove this line: https://github.com/orbiting/backends/blob/master/servers/republik/modules/crowdfundings/graphql/schema.js#L66

**Comment by tpreusse** [comment]
> @tpreusse could we then remove this line: https://github.com/orbiting/backends/blob/master/servers/republik/modules/crowdfundings/graphql/schema.js#L66

No because [republik-admin-frontend](https://github.com/orbiting/republik-admin-frontend) does use it.

**Comment by caillou** [comment]
@patte I just deployed the current state and did a manual test. Could you review one more time, just in case?

File: servers/republik/modules/crowdfundings/graphql/resolvers/Membership.js [hunk 2]
```diff
@@ -29,19 +29,25 @@ const membershipResolver = {
...
 
-    const latest = await pgdb.public.membershipPeriods.findFirst({
-      membershipId: membership.id
-    }, { fields: '"endDate"', orderBy: ['endDate desc'] })
-    return !!(
-      membership.active &&
-      membership.latestPaymentFailedAt &&
-      membership.latestPaymentFailedAt > latest.endDate
-    )
+    const latestPeriod = periods[periods.length - 1]
```
Comment (patte) [inline servers/republik/modules/crowdfundings/graphql/resolvers/Membership.js:46]:
```suggestion
    const latestPeriod = periods[0]
```
periods is sorted `DESC`.

File: servers/republik/modules/crowdfundings/graphql/resolvers/Membership.js [hunk 3]
```diff
@@ -29,19 +29,25 @@ const membershipResolver = {
...
+    const isLatestPeriodEnded = new Date(latestPeriod.endDate) < new Date()
+    return isLatestPeriodEnded
```
Comment (patte) [inline servers/republik/modules/crowdfundings/graphql/resolvers/Membership.js:48]:
```suggestion
    return new Date(latestPeriod.endDate) < new Date()
```

File: servers/republik/modules/crowdfundings/lib/cancelMembership.js [hunk 4]
```diff
@@ -2,9 +2,15 @@ const debug = require('debug')('crowdfundings:cancelMembership')
...
 const cancelSubscription = require('./payments/stripe/cancelSubscription')
 
+const { overdue } = require('../graphql/resolvers/Membership')
+
 module.exports = async (membership, details, options, t, pgdb) => {
   const { reason, type, suppressConfirmation, suppressWinback, cancelledViaSupport } = details
-  const { immediately } = options
+  let { immediately } = options
+
+  if (overdue(membership, null, { pgdb })) {
```
Comment (patte) [inline servers/republik/modules/crowdfundings/lib/cancelMembership.js:11]:
```suggestion
  if (!immediately && await overdue(membership, null, { pgdb })) {
```
:point_up:

**Pull request merged**